osquery/tables/system/darwin/keychain_utils.cpp

```diff
@@ -72,151 +73,132 @@ std::string getKeychainPath(const SecKeychainItemRef& item) {
   return path;
 }
 
-std::string genKIDProperty(const CFDataRef& kid) {
-  CFDataRef kid_data = nullptr;
-  CFDictionaryRef kid_dict = nullptr;
-
-  // Find the key identifier data within the property mess.
-  for (CFIndex i = 0; i < CFArrayGetCount((CFArrayRef)kid); i++) {
-    kid_dict = (CFDictionaryRef)CFArrayGetValueAtIndex((CFArrayRef)kid, i);
-    auto kid_value =
-        (const char*)CFDictionaryGetValue(kid_dict, kSecPropertyKeyValue);
-
-    if (CFGetTypeID(kid_value) == CFDataGetTypeID()) {
-      kid_data = (CFDataRef)kid_value;
-      break;
-    }
+std::string genKIDProperty(const unsigned char* data, int len) {
+  std::stringstream key_id;
+  for (int i = 0; i < len; i++) {
+    key_id << std::setw(2) << std::hex << std::setfill('0') << (int)data[i];
   }
+  return key_id.str();
+}
 
-  if (kid_data == nullptr) {
-    // No key identifier found.
+std::string genAlgProperty(const X509* cert) {
+  int pub_key_alg_nid = OBJ_obj2nid(cert->cert_info->key->algor->algorithm);
+  if (pub_key_alg_nid == NID_undef) {
+    // Unknown algorithm OID.
     return "";
   }
+  return std::string(OBJ_nid2ln(pub_key_alg_nid));
+}
 
-  // Provide an ASCII-representation of the KID, similar to keychain.
-  std::stringstream ascii_kid;
-  for (CFIndex i = 0; i < CFDataGetLength(kid_data); i++) {
-    int kid_byte = (uint8_t)CFDataGetBytePtr(kid_data)[i];
-    ascii_kid << std::setfill('0') << std::setw(2) << std::hex << kid_byte;
-    // Then make it easy to read.
-    if (i < CFDataGetLength(kid_data) - 1) {
-      ascii_kid << "";
-    }
-  }
-
-  return ascii_kid.str();
+std::string genSHA1ForCertificate(const CFDataRef& raw_cert) {
+  return hashFromBuffer(
+      HASH_TYPE_SHA1, CFDataGetBytePtr(raw_cert), CFDataGetLength(raw_cert));
 }
 
-std::string genCommonNameProperty(const CFDataRef& ca) {
-  CFDataRef ca_data = nullptr;
-  CFStringRef ca_string = nullptr;
+bool CertificateIsCA(X509* cert) {
+  // The following codes are returned by X509_check_ca():
+  // 0 => not a CA, 1 => is a CA, 2 => basicConstraints absent so "maybe" a CA,
+  // 3 => basicConstraints absent but self signed V1,
+  // 4 => basicConstraints absent but keyUsage present and keyCertSign asserted.
+  // Some applications treat >= 1 as CA, but OS X doesn't, which is what we do.
+  return (X509_check_ca(cert) == 1);
```

Comment:
Yup, this function should accept more return codes. On 10.10, Keychain Access seems to mark certificates as CA's when `X509_check_ca` returns 1, 3, or 5.

To find that out, I made [a little utility](https://gist.github.com/mark-ignacio/b9015b681630552cdfad) to run the entirety of OS X's CA store against `X509_check_ca`, and I found a bunch of 1's, 3's, and some expired 5's.

I don't know how to easily craft a certificate that makes it output 4 to test it, but the source for `X509_check_ca` used in 10.10 is here:  http://opensource.apple.com/source/OpenSSL098/OpenSSL098-52/src/crypto/x509v3/v3_purp.c. 

It also doesn't seem return a 2 or 4, so I don't know what those comments about codes 2 and 4 are about...?